src/services/download/mod.rs

```diff
@@ -85,43 +85,42 @@ impl DownloadService {
         source: SourceFileId,
         destination: PathBuf,
     ) -> Box<dyn Future<Item = DownloadStatus, Error = DownloadError> + Send + 'static> {
-        let fut03 = self.worker.spawn(|| async move {
-            match source {
-                SourceFileId::Sentry(source, loc) => {
-                    sentry::download_source(source, loc, destination)
-                        .compat()
-                        .await
-                }
-                SourceFileId::Http(source, loc) => {
-                    http::download_source(source, loc, destination)
-                        .compat()
-                        .await
-                }
-                SourceFileId::S3(source, loc) => {
-                    s3::download_source(source, loc, destination).compat().await
-                }
-                SourceFileId::Gcs(source, loc) => {
-                    gcs::download_source(source, loc, destination)
-                        .compat()
-                        .await
-                }
-                SourceFileId::Filesystem(source, loc) => {
-                    filesystem::download_source(source, loc, destination)
+        let fut03 = self.worker.spawn(
+            "service.download",
+            std::time::Duration::from_secs(3600),
+            || async move {
+                match source {
+                    SourceFileId::Sentry(source, loc) => {
+                        sentry::download_source(source, loc, destination)
+                            .compat()
+                            .await
+                    }
+                    SourceFileId::Http(source, loc) => {
+                        http::download_source(source, loc, destination)
+                            .compat()
+                            .await
+                    }
+                    SourceFileId::S3(source, loc) => {
+                        s3::download_source(source, loc, destination).compat().await
+                    }
+                    SourceFileId::Gcs(source, loc) => {
+                        gcs::download_source(source, loc, destination)
+                            .compat()
+                            .await
+                    }
+                    SourceFileId::Filesystem(source, loc) => {
+                        filesystem::download_source(source, loc, destination)
+                    }
                 }
-            }
-        });
+            },
+        );
+
+        // Map all SpawnError variants into DownloadErrorKind::Canceled.
         let fut01 = fut03
             .map(|spawn_ret| spawn_ret.unwrap_or_else(|_| Err(DownloadErrorKind::Canceled.into())))
             .boxed()
             .compat();
-        Box::new(future_metrics!(
-            "service.download",
-            Some((
-                std::time::Duration::from_secs(3600),
-                DownloadErrorKind::Canceled.into()
-            )),
-            fut01
-        ))
+        Box::new(fut01)
```

Comment:
you have `Box::new(.boxed().compat())`, can you reorder this in order to avoid the double boxing?

Reply:
IIRC no.  The `TryFuture` trait (which is where this `.compat()` comes from) is implemented for `std::future::Future` but the `.map()` returns a `futures::future::Map` and we need to box to make it a `TryFuture`.

Note that this goes away in #241